File: lua/strain.lua
local function keep(xs, pred)
    local t = {}
    for _, x in ipairs(xs) do
        if pred(x) then
            table.insert(t, x)
        end
    end
    return t
end

local function discard(xs, pred)
    p = function(x)
        return not pred(x)
    end
    return keep(xs, p)
end

return { keep = keep, discard = discard }

File: go/strain.go
package strain

func Keep[T any](s []T, fn func(T) bool) []T {
	t := []T{}
	for _, each := range s {
		if fn(each) {
			t = append(t, each)
		}
	}
	return t
}

func Discard[T any](s []T, fn func(T) bool) []T {
	t := []T{}
	for _, each := range s {
		if !fn(each) {
			t = append(t, each)
		}
	}
	return t
}
